### PR TITLE
Multi zones

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---
Allow multi zones in dvectors

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

File: src/caf/base/data_structures.py [hunk 1]
```diff
@@ -2226,7 +2470,13 @@ def balance_by_segments(
         )
 
 
-@dataclass
+class Config:
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2473]: Make this private e.g. `_Config`

File: src/caf/base/data_structures.py [hunk 2]
```diff
@@ -2256,6 +2506,14 @@ class IpfTarget:
         None  # keys are segment in target, values, segment in seed
     )
 
+    @pydantic.model_validator(mode="after")
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2509]: If using mode=after then the method is an instance method so self can be used

File: src/caf/base/segments/m_nhb.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Segment config written on 2024-06-28 at 11:07
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/segments/m_nhb.yml:1]: See comment on m_hb

File: src/caf/base/data_structures.py [hunk 4]
```diff
@@ -1108,50 +1309,67 @@ def split_by_other(self, other: DVector, agg_zone: ZoningSystem | None = None):
...
                 "please use the expand_to_other method."
             )
+
+        common = list(self.segmentation.overlap(other.segmentation))
+        check = False
+        if isinstance(other.zoning_system, Sequence):
+            assert isinstance(other.zoning_system, Sequence)
+            check = agg_zone in other.zoning_system
+            if isinstance(self.zoning_system, Sequence):
+                assert isinstance(self.zoning_system, Sequence)
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:1319]: I am 99% certain these are just for mypy yes

File: src/caf/base/data_structures.py [hunk 5]
```diff
@@ -1305,12 +1523,16 @@ def combine_from_dic(
         comb = {val: dvec.data for val, dvec in in_dic.items()}
         new_data = pd.concat(comb)
         new_segmentation = in_segmentation.add_segment(new_seg)
+        new_data.index.names = list(
+            map(lambda x: new_seg.name if x is None else x, new_data.index.names)
+        )
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:1528]: No, all index levels of the data within a DVector are named

File: src/caf/base/data_structures.py [hunk 6]
```diff
@@ -1519,6 +1751,7 @@ def calc_rmse(self, targets: list[IpfTarget]) -> float:
                     raise TypeError(
                         "A translation is provided but the target has no zoning_system."
                     )
+                assert isinstance(target.data.zoning_system, ZoningSystem)
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:1754]: This is wrong but addressed in another PR. IPFtargets can't have multizones but they *can* have no zone system

File: src/caf/base/data_structures.py (hunk 2 above)
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:2509]: So this doesn't need to be a class method?

File: src/caf/base/segments.py [hunk 7]
```diff
@@ -232,15 +238,19 @@ class SegmentsSuper(enum.Enum):
     """
 
     PURPOSE = "p"
+    PURPOSE_HB = "p_hb"
+    PURPOSE_NHB = "p_nhb"
     TIMEPERIOD = "tp"
     MODE = "m"
+    MODE_HB = "m_hb"
+    MODE_NHB = "m_nhb"
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/segments.py:246]: Addressed elsewhere - keeping as is for now will open another PR to sort out duplicate segments

File: RELEASE.md [hunk 8]
```diff
@@ -20,4 +20,6 @@ there.
 Below, a brief summary of patches made since the previous version can be found.
 
 ### V0.1.0
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply RELEASE.md:22]: So this file can be deleted completely?

File: src/caf/base/segments/m_nhb.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/base/segments/m_nhb.yml:1]: I can't see a comment

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Looks like there are a few of previous comments which were missed.

I think it's mainly:
- a couple of docstrings,
- purpose 17 missing from direction_od exclusion
- RELEASE.md can be deleted

Also, the new release will have some API changs mainly data types for `DVector.zoning_system`, so should be a MINOR bump.

File: src/caf/base/data_structures.py [hunk 9]
```diff
@@ -247,7 +250,7 @@ def __init__(
         self,
         segmentation: Segmentation,
         import_data: pd.DataFrame,
-        zoning_system: Optional[ZoningSystem] = None,
+        zoning_system: Optional[ZoningSystem | Sequence[ZoningSystem]] = None,
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:251]: Sorry didn't notice this comment updating now
